Tests/OpenAPIRuntimeTests/Base/Test_MIMEType.swift

```diff
@@ -0,0 +1,102 @@
+//
+// This source file is part of the SwiftOpenAPIGenerator open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftOpenAPIGenerator project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftOpenAPIGenerator project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import XCTest
+@_spi(Generated) import OpenAPIRuntime
+
+final class Test_MIMEType: Test_Runtime {
+    func test() throws {
+        let cases: [(String, MIMEType?, String?)] = [
+
+            // Common
+            (
+                "application/json",
+                MIMEType(
+                    type: .concrete("application"),
+                    subtype: .concrete("json")
+                ),
+                "application/json"
+            ),
+
+            // Subtype wildcard
+            (
+                "application/*",
+                MIMEType(
+                    type: .concrete("application"),
+                    subtype: .wildcard
+                ),
+                "application/*"
+            ),
+
+            // Type wildcard
+            (
+                "*/*",
+                MIMEType(
+                    type: .wildcard,
+                    subtype: .wildcard
+                ),
+                "*/*"
+            ),
+
+            // Common with a parameter
+            (
+                "application/json; charset=UTF-8",
+                MIMEType(
+                    type: .concrete("application"),
+                    subtype: .concrete("json"),
+                    parameters: [
+                        "charset": "UTF-8"
+                    ]
+                ),
+                "application/json; charset=UTF-8"
+            ),
+
+            // Common with two parameters
+            (
+                "application/json; charset=UTF-8; boundary=1234",
+                MIMEType(
+                    type: .concrete("application"),
+                    subtype: .concrete("json"),
+                    parameters: [
+                        "charset": "UTF-8",
+                        "boundary": "1234",
+                    ]
+                ),
+                "application/json; boundary=1234; charset=UTF-8"
+            ),
+
+            // Common case preserving, but case insensitive equality
+            (
+                "APPLICATION/JSON;CHARSET=UTF-8",
+                MIMEType(
+                    type: .concrete("application"),
+                    subtype: .concrete("json"),
+                    parameters: [
+                        "charset": "UTF-8"
+                    ]
+                ),
+                "APPLICATION/JSON; CHARSET=UTF-8"
+            ),
+
+            // Invalid
+            ("application", nil, nil),
+            ("application/foo/bar", nil, nil),
+            ("", nil, nil),
```

Comment:
Is it possible to construct `*/json` currently? If so, should it be possible?

Reply:
By my reading of https://datatracker.ietf.org/doc/html/rfc2045#section-5.1, that wouldn't be a legal MIME type. We wouldn't reject it at parsing though today, it'd just never positively match any provided MIME type. Should we reject it in parsing you think?

Reply:
I would argue we probably don't even want to allow creating invalid `MIMETypes`, so making the `init` `throws` and adding a check there would be what I'd go for.

Reply:
Today we have a failable initializer `init?(_ description: String)` which allows us to reject invalid inputs. We don't try to return _why_ it's invalid, we just return nil if it's invalid.

Maybe the type should reflect the limit in the type system? Like

```swift
struct MIMEType {
  enum Kind {
    case any
    case anySubtype(type: String)
    case concrete(type: String, subtype: String)
  }
  var kind: Kind
  var parameters: [String: String]
}
```

WDYT?

Reply:
I'm happy with either `init?` or `init throws`. And yes, I think pushing this expressivity into the type system is even better.

But as you say, it's SPI, so I'm not precious over  this.

Reply:
I'm divided on this - I somewhat like it, but you could still create a `concrete` type with wildcards (or even worse, something invalid like `.concrete(type: "*", subtype: "json")`), so I'm not sure whether the extra complexity is worth it.

Reply:
I couldn't unsee it, so I redid it, and like this much better: https://github.com/apple/swift-openapi-runtime/pull/29/commits/de2248b6beecc256d2b5be5c74fcd1c849e0f4db